internal/controllers/common_controller.go

```diff
@@ -26,6 +27,31 @@ import (
 	ocv1alpha1 "github.com/operator-framework/operator-controller/api/v1alpha1"
 )
 
+// setInstalledStatusFromBundle sets the installed status based on the given installedBundle.
+func setInstalledStatusFromBundle(ext *ocv1alpha1.ClusterExtension, installedBundle *InstalledBundle, err error) {
+	// Something is installed
+	if installedBundle != nil {
+		installStatus := &ocv1alpha1.ClusterExtensionInstallStatus{
+			Bundle: installedBundle.BundleMetadata,
+		}
+		setInstallStatus(ext, installStatus)
+		// If both conditions occur, this is a failed Apply when a bundle was already installed
+		if err != nil {
+			setInstalledStatusConditionFailed(ext, err.Error())
```

Comment:
Wouldn't we want to preserve the historical installation information on a failed "upgrade" attempt, which is what I interpret a "failed Apply when a bundle was already installed" to mean.

Reply:
I was concerned about this, but the problem is that we lose the fact that we failed an upgrade. I'm not sure which is better. But this is what @joelanford suggested.

Reply:
Another way to look at it: if we tried an install, we store the bundle information. Should we not have the same information on an upgrade?
I almost think we need to have an upgrade field...

Reply:
There's additional discussion here: https://github.com/operator-framework/operator-controller/pull/1379#discussion_r1805660826

Reply:
Also note, that this change (no longer preserving `Installed` after the `Apply()`) is a separate fixup commit, so it can be looked at in isolation.

Reply:
If a ClusterExtension has been installed successfully in the past, we should keep the installed condition as is and include the upgrade failure information in the progressing condition. The resolution information can be included in the error message so that a user can better connect the dots that an upgrade failed and what bundles were involved.